File: prob23.py
def is_abundant(n):
	sum = 1
	for i in range(2, n):
		if n % i == 0:
			sum = sum + i
	return sum > n


abundant = [12]

max_not_two_ab = 28123;

for i in range(13,28123-12+1):
	if is_abundant(i):
		abundant.append(i)

sums = []
for i in range(0, len(abundant)):
	for x in range(i,len(abundant)):
		sums.append(i + x)
sum = 0
for x in range(24, 28124):
	if x in sums:
		sum = sum + x

print sum

File: prob23.scala

object prob23 {
  def main(args: Array[String]): Unit = {
    val abundants:Set[Int] = (1 to 30100).filter(isAbundant).toSet
    val abundantSums:Set[Int] = abundants.subsets(2).map(_.sum).toSet ++ abundants.map(_*2)
    println((1 to 30200).filterNot(abundantSums.contains).sum)
  }

  def isAbundant(i: Int) : Boolean = {
    getFactors(i).sum > i
  }

  def getFactors(input: Int) : List[Int] = {
    val primeFactors = getPrimeFactors(input)
    primeFactors.indices.flatMap(primeFactors.combinations).map(_.product).toList
  }

  def getPrimeFactors(input: Int): List[Int] =  {
    var x = input
    var out:List[Int] = List()
    var divider = 2
    while (x > 1) {
      if (x % divider == 0) {
        out = divider :: out
        x /= divider
      } else {
        divider += 1
      }
    }
    out
  }
}
